### ISSUE TITLE
Jellyfin needs to convert 10-bit to 8-bit when the encoder does not support 10-bit encoding

### ISSUE THREAD

**Issue by barolo** [jellyfin/jellyfin#1527]
**Describe the bug**

10bit HEVC files fail transcode resulting in infinite loading animation upon opening video.

Hardware AMD Ryzen 3 2200G APU

**Logs**

[full debug log](https://lnk.skro.pl/9) from jellyfin ffmpeg command [](url)

VAINFO:
libva info: VA-API version 1.4.1
libva info: va_getDriverName() returns 0
libva info: Trying to open /usr/lib64/va/drivers/radeonsi_drv_video.so
libva info: Found init function __vaDriverInit_1_4
libva info: va_openDriver() returns 0
vainfo: VA-API version: 1.4 (libva 2.4.0)
vainfo: Driver version: Mesa Gallium driver 19.2.0-devel for AMD RAVEN (DRM 3.30.0, 5.1.10-gentoo, LLVM 8.0.0)
vainfo: Supported profile and entrypoints
      VAProfileMPEG2Simple            : VAEntrypointVLD
      VAProfileMPEG2Main              : VAEntrypointVLD
      VAProfileVC1Simple              : VAEntrypointVLD
      VAProfileVC1Main                : VAEntrypointVLD
      VAProfileVC1Advanced            : VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointVLD
      VAProfileH264ConstrainedBaseline: VAEntrypointEncSlice
      VAProfileH264Main               : VAEntrypointVLD
      VAProfileH264Main               : VAEntrypointEncSlice
      VAProfileH264High               : VAEntrypointVLD
      VAProfileH264High               : VAEntrypointEncSlice
      VAProfileHEVCMain               : VAEntrypointVLD
      VAProfileHEVCMain               : VAEntrypointEncSlice
      VAProfileHEVCMain10             : VAEntrypointVLD
      VAProfileJPEGBaseline           : VAEntrypointVLD
      VAProfileVP9Profile0            : VAEntrypointVLD
      VAProfileVP9Profile2            : VAEntrypointVLD
      VAProfileNone                   : VAEntrypointVideoProc

ffmpeg  -hide_banner -encoders | grep vaapi
 V..... h264_vaapi           H.264/AVC (VAAPI) (codec h264)
 V..... hevc_vaapi           H.265/HEVC (VAAPI) (codec hevc)
 V..... mjpeg_vaapi          MJPEG (VAAPI) (codec mjpeg)
 V..... mpeg2_vaapi          MPEG-2 (VAAPI) (codec mpeg2video)
 V..... vp8_vaapi            VP8 (VAAPI) (codec vp8)
 V..... vp9_vaapi            VP9 (VAAPI) (codec vp9)

**System (please complete the following information):**
 - OS: Gentoo
 - Browser: any
 - Jellyfin Version: 10.3.6

MPV is able to play these files just fine with vaapi accel and low CPU usage [ 3% ]
Not sure if it's helpful but I'm attaching how it looks in mpv with same file and working accell, [debug log](https://lnk.skro.pl/a)

**Comment by barolo** [comment]
Update, it's a mesa bug, in part. Patch which fixes it is attached [here](https://trac.ffmpeg.org/ticket/8006).
But the issue still stand as jellyfish ffmpeg command doesn't work, as it needs to convert 10 -> 8Bit.

Something like this does:
ffmpeg -threads 4 \
-init_hw_device vaapi=amd:/dev/dri/renderD128 -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device amd -filter_hw_device amd \
-i trk.mkv  \
-vf "scale_vaapi=format=nv12" \
-c:v hevc_vaapi \
-c:a copy  -c:s copy \
-y plop.mkv

**Comment by EraYaN** [comment]
It's quite interesting that the hardware does not do h264 in 10 bit. I'll edit the title.

**Comment by barolo** [comment]
> It's quite interesting that the hardware does not do h264 in 10 bit. I'll edit the title.
> 
> This might be more difficult to implement that you expect, ffmpeg will report certain formats as supported pixel formats, even if they are not by the underlying hardware. At worst this means admins will manually have to put stuff into the settings.

Yeah, thank you, I'm aware of the issues, e.g my hardware supports vp9 and vainfo reports vp9 encode but it's still not implemented so ffmpeg will fail.
That said hevc 10bit seems like the more important one, and is gaining popularity fast due to quality/space savings ratio, especially important for a big media libraries

**Comment by EraYaN** [comment]
Right but when clients support HEVC, this issue does not show. And if we ever add the HEVC encode as well, this should also not be an issue. Pixel format conversion is more complex then it might seem, especially with hardware encoding. But if anyone want to give it a go, by all means.

**Comment by barolo** [comment]
I have client which supports hevc but not 10 bit, weirdly. 
I've seen this suggested already; it would make things easier for devs probably if users were able to force own per device ffmpeg commands, detection is very flaky atm anyway

**Comment by barolo** [comment]
After some digging figured out what proper commands for conversions 10bit->8bit  should like, results are surprisingly decent for 4x speed hardware transcode.

hevc->h264
```
fmpeg  -init_hw_device vaapi=amd:/dev/dri/renderD128 \
-hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device amd -filter_hw_device amd \
-i nexp.mkv \
-vf "hwdownload,format=p010le,format=nv12,hwupload" -c:v h264_vaapi -profile:v 578  \
-threads 4 -c:a copy -c:s copy -f mpegts plop.mkv

```
hevc10bit->hevc8bit 
```
fmpeg  -init_hw_device vaapi=amd:/dev/dri/renderD128 \
-hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device amd -filter_hw_device amd \
-i nexp.mkv \
-vf "hwdownload,format=p010le,format=nv12,hwupload" hevc_vaapi -profile:v main   \
-threads 4 -c:a copy -c:s copy -f mpegts plop.mkv
```

**Comment by oddstr13** [comment]
Is that `hevc_vaapi` missing a `-c:v` ?

**Comment by barolo** [comment]
> Is that `hevc_vaapi` missing a `-c:v` ?

Indeed, thanks!

**Comment by barolo** [comment]
Closed my mistake

**Comment by barolo** [comment]
@EraYaN I'm not sure if the title reflects what the actual issue is, I don't want to encode 10 bit, none of my mobile clients can handle 10 bit even if it is supported and can sluggishly play it. I want to transcode 10 bit to 8 bit, be it to hevc or x264, and that's where jelly fails, in pixel format conversion.
I'm talking about hardware accelerated transcoding, my APU doesn't have 10 bit hardware encoders anyway [ only newest stuff does? ]

**Comment by scurrvy2020** [comment]
> After some digging figured out what proper commands for conversions 10bit->8bit should look like, at least on modern amd, results are surprisingly decent for 4x speed hardware transcode.
> 
> hevc_10bit->h264_8bit
> 
> ```
> ffmpeg  -init_hw_device vaapi=amd:/dev/dri/renderD128 \
> -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device amd -filter_hw_device amd \
> -i nexp.mkv \
> -vf "hwdownload,format=p010le,format=nv12,hwupload" -c:v h264_vaapi -profile:v 578  \
> -threads 4 -c:a copy -c:s copy -f mpegts plop.mkv
> ```
> 
> hevc_10bit->hevc_8bit
> 
> ```
> ffmpeg  -init_hw_device vaapi=amd:/dev/dri/renderD128 \
> -hwaccel vaapi -hwaccel_output_format vaapi -hwaccel_device amd -filter_hw_device amd \
> -i nexp.mkv \
> -vf "hwdownload,format=p010le,format=nv12,hwupload" -c:v hevc_vaapi -profile:v main   \
> -threads 4 -c:a copy -c:s copy -f mpegts plop.mkv
> ```

Is there anywhere we can put these commands to make it work?

**Comment by sammyrc34** [comment]
I'm making a change to partially fix this in the case of VAAPI decoding with software encoding. I didn't do anything to hardware VAAPI encoding as 10bit VAAPI pixel format conversions are broken in the included ffmpeg 4.0. See #1675.

**Comment by barolo** [comment]
> I'm making a change to partially fix this in the case of VAAPI decoding with software encoding. I 
@sammyrc34 

didn't do anything to hardware VAAPI encoding as 10bit VAAPI pixel format conversions are broken in the included ffmpeg 4.0. See #1675.

Aren't hey moving to 4.1 and 4.2 currently? There was one bug with skipping which got fixed afaik

> EDIT: Also, currently capable pixel formats aren't configurable for clients. To really deal with this problem properly, I think we need to have pixel formats and codecs configurable. There are devices that support HEVC 8-bit and H264 10-bit for example.

I have one such device, that's what prompted the whole thing, I also have one device which supports 10 bit fully but chokes on bitrate

**Comment by sammyrc34** [comment]
@barolo 

> I have one such device, that's what prompted the whole thing, I also have one device which supports 10 bit fully but chokes on bitrate

That's a perfect argument to enable pixel format configuration for clients. If I find more time later, I might look at implementing such a change.

> hevc software encoding on my apu is way too slow to be usable for streaming, but getting vaapi decoding would be nice

Agreed.

**Comment by joshuaboniface** [comment]
Sorry for the major delay - yes, we're moving to FFMpeg 4.2.1 for 10.4.0, which should include 10-bit support AFAIK.

**Comment by barolo** [comment]
> Sorry for the major delay - yes, we're moving to FFMpeg 4.2.1 for 10.4.0, which should include 10-bit support AFAIK.

@joshuaboniface I've been testing those issues with 4.2+ for quite some time, last checked, for AMD it still needs proper commands for pixel format conversion.
Will recheck what's the state of it now.

**Comment by barolo** [comment]
Unsure if it's worth pursuing, there's weird hostility on ffmpeg/mesa vaapi trackers and everything moves at snails pace in related space. Supposedly Linux platform will get amf encoding early next year [It's using proprietary vulkan extensions so it's limited to closed source drivers for now,  till khronos finalizes encoding bits for vulkan ], some bits and pieces already landed in ffmpeg, looks like  vendor specific encoding is the way to go.
AMF is also quite a bit faster from  my Window tests.

**Comment by barolo** [comment]
This issue got fixed with a patch for #2473 [ at least with mesa git ] just tested it, works very well.
After inclusion in mesa I'm going to close this.

**Comment by barolo** [comment]
Needed patch just got merged into mesa-git, thus I'm closing this.
